File: lab6-openmp/lab6-openmp_cxx/omp_orphan.cpp
/* FILE: omp_orphan.cpp
   DESCRIPTION:
      OpenMP Example - Parallel region with an orphaned directive - C++ Version 
      This example demonstrates a dot product being performed by an orphaned
      loop reduction construct.  Scoping of the reduction variable is critical.
   AUTHOR: Blaise Barney  5/99
   UPDATED: Daniel R. Reynolds (updated to C++), 1/13/2013 */

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>
#define VECLEN 100

// global variables
float a[VECLEN], b[VECLEN], sum;

// external function
void dotprod() {

  // local variables 
  int i, tid;

  tid = omp_get_thread_num();
# pragma omp for reduction(+:sum)
  for (i=0; i < VECLEN; i++) {
    sum = sum + (a[i]*b[i]);
    printf("  tid = %i,  i = %i\n", tid, i);
  }
}


// main routine
int main (int argc, char *argv[]) {

  // local variables
  int i;

  // initialize values
  for (i=0; i<VECLEN; i++)   a[i] = b[i] = 1.0 * i;
  sum = 0.0;

  // begin parallel region, with orphaned call to dotprod()
# pragma omp parallel
  dotprod();

  // output result
  printf("Sum = %g\n", sum);

  return 0;
} // end main

File: lab6-openmp/lab6-openmp_f90/omp_orphan.f90
! -*- Mode: Fortran90; -*-
!=================================================================

program Orphan
  !-----------------------------------------------------------------
  ! FILE: omp_orphan.f90
  ! DESCRIPTION:
  !   OpenMP Example - Parallel region with an orphaned directive
  !   This example demonstrates a dot product being performed by 
  !   an orphaned loop reduction construct.  Scoping of the 
  !   reduction variable is critical.
  ! AUTHOR: Blaise Barney  5/99
  ! Upgraded to F90, free-source form: Daniel R. Reynolds, 2/2009
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer :: i
  integer, parameter :: N = 100
  real*8 :: A(N), B(N), sum

  !======= Internals ============

  do i=1, N
     A(i) = 1.d0 * i
     B(i) = A(i)
  enddo
  sum = 0.d0

  !$omp parallel
  call DotProd(A, B, sum)
  !$omp end parallel

  write(*,*) "Sum = ", sum

end program Orphan
!=================================================================



subroutine DotProd(A, B, sum)
  !-----------------------------------------------------------------
  ! DOTPROD: Computes the dot-product and places the result in SUM.
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer :: i, TID
  integer, external :: omp_get_thread_num
  integer, parameter :: N = 100
  real*8 :: A(N), B(N), sum

  !======= Internals ============

  TID = omp_get_thread_num()
  !$omp do reduction(+:sum)
  do i=1, N
     sum = sum + (A(i)*B(i))
     print *, '  TID= ',TID,'I= ',i
  enddo
  !$omp end do

  return

end subroutine DotProd
!=================================================================
